File: ios/nativebrik_bridge/Sources/nativebrik_bridge/Manager.swift
//
//  Manager.swift
//  nativebrik_bridge
//
//  Created by Ryosuke Suzuki on 2024/03/10.
//

import Foundation
import Flutter
import UIKit
@_spi(FlutterBridge) import Nubrick

struct EmbeddingEntity {
    let uiview: UIView
    let channel: FlutterMethodChannel
    let accessor: __DO_NOT_USE__NativebrikBridgedViewAccessor?
}

struct RemoteConfigEntity {
    let variant: RemoteConfigVariant?
}

class NativebrikBridgeManager {
    private var nubrickClient: NubrickClient? = nil
    private var embeddingMaps: [String:EmbeddingEntity]
    private var configMaps: [String:RemoteConfigEntity]

    init() {
        self.embeddingMaps = [:]
        self.configMaps = [:]
    }

    func setNativebrikClient(nativebrik: NubrickClient) {
        if self.nubrickClient != nil {
            return print("NativebrikClient is already set")
        }
        self.nubrickClient = nativebrik
        if let vc = UIApplication.shared.delegate?.window??.rootViewController {
            let overlay = nativebrik.experiment.overlayViewController()
            vc.addChild(overlay)
            vc.view.addSubview(overlay.view)
        }
    }

    func getUserId() -> String? {
        guard let nubrickClient = self.nubrickClient else {
            return nil
        }
        return nubrickClient.user.id
    }

    func setUserProperties(properties: [String: Any]) {
        guard let nubrickClient = self.nubrickClient else {
            return
        }
        nubrickClient.user.setProperties(properties)
    }

    func getUserProperties() -> [String: String]? {
        guard let nubrickClient = self.nubrickClient else {
            return nil
        }
        return nubrickClient.user.getProperties()
    }

    // embedding
    func connectEmbedding(id: String, channelId: String, arguments: Any?, messenger: FlutterBinaryMessenger) {
        guard let nubrickClient = self.nubrickClient else {
            return
        }
        let channel = FlutterMethodChannel(name: "Nativebrik/Embedding/\(channelId)", binaryMessenger: messenger)
        let uiview = nubrickClient.experiment.embeddingForFlutterBridge(id, arguments: arguments, onEvent: { event in
            channel.invokeMethod(ON_EVENT_METHOD, arguments: [
                "name": event.name as Any?,
                "deepLink": event.deepLink as Any?,
                "payload": event.payload?.map({ prop in
                    return [
                        "name": prop.name,
                        "value": prop.value,
                        "type": prop.type
                    ]
                }),
            ])
        }, onSizeChange: { width, height in
            channel.invokeMethod(EMBEDDING_SIZE_UPDATE_METHOD, arguments: [
                "width": width,
                "height": height,
            ])
        }) { phase in
            switch phase {
            case .completed(let view):
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "completed")
                return view
            case .notFound:
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "not-found")
                return UIView()
            case .failed:
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "failed")
                return UIView()
            case .loading:
                return UIView()
            }
        }
        let embeedingEntity = EmbeddingEntity(
            uiview: uiview,
            channel: channel,
            accessor: nil
        )
        self.embeddingMaps[channelId] = embeedingEntity
    }

    func disconnectEmbedding(channelId: String) {
        self.embeddingMaps[channelId] = nil
    }

    func getEmbeddingEntity(channelId: String) -> EmbeddingEntity? {
        guard let entity = self.embeddingMaps[channelId] else {
            return nil
        }
        return entity
    }

    // remote config
    func connectRemoteConfig(id: String, channelId: String, onPhase:  @escaping ((RemoteConfigPhase) -> Void)) {
        guard let nubrickClient = self.nubrickClient else {
            return
        }
        let entity = RemoteConfigEntity(variant: nil)
        self.configMaps[channelId] = entity

        nubrickClient.experiment.remoteConfig(id) { phase in
            switch phase {
            case .completed(let config):
                if self.configMaps[channelId] == nil {
                    // disconnected already
                    return
                }
                let entity = RemoteConfigEntity(variant: config)
                self.configMaps[channelId] = entity
                onPhase(phase)
            case .notFound:
                onPhase(phase)
            case .failed:
                onPhase(phase)
            default:
                break
            }
        }
    }

    func disconnectRemoteConfig(channelId: String) {
        self.configMaps[channelId] = nil
    }

    func connectEmbeddingInRemoteConfigValue(key: String, channelId: String, arguments: Any?, embeddingChannelId: String, messenger: FlutterBinaryMessenger) {
        guard let entity = self.configMaps[channelId] else {
            return
        }
        guard let variant = entity.variant else {
            return
        }
        let channel = FlutterMethodChannel(name: "Nativebrik/Embedding/\(embeddingChannelId)", binaryMessenger: messenger)
        guard let uiview = variant.getAsUIView(key, arguments: arguments, onEvent: { event in
            channel.invokeMethod(ON_EVENT_METHOD, arguments: [
                "name": event.name as Any?,
                "deepLink": event.deepLink as Any?,
                "payload": event.payload?.map({ prop in
                    return [
                        "name": prop.name,
                        "value": prop.value,
                        "type": prop.type
                    ]
                }),
            ])
        }, content: { phase in
            switch phase {
            case .completed(let view):
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "completed")
                return view
            case .notFound:
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "not-found")
                return UIView()
            case .failed:
                channel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, arguments: "failed")
                return UIView()
            case .loading:
                return UIView()
            }
        }) else {
            return
        }
        let embeedingEntity = EmbeddingEntity(
            uiview: uiview,
            channel: channel,
            accessor: nil
        )
        self.embeddingMaps[embeddingChannelId] = embeedingEntity
    }

    func getRemoteConfigValue(channelId: String, key: String) -> String? {
        guard let entity = self.configMaps[channelId] else {
            return nil
        }
        guard let variant = entity.variant else {
            return nil
        }
        return variant.get(key)
    }

    // tooltip
    func connectTooltip(name: String, onFetch: @escaping (String) -> Void, onError: @escaping (String) -> Void) {
        guard let nubrickClient = self.nubrickClient else {
            onError("NativebrikClient is not set")
            return
        }
        Task {
            let result = await nubrickClient.experiment.__do_not_use__fetch_tooltip_data(trigger: name)
            switch result {
            case .success(let data):
                onFetch(data)
            case .failure(let error):
                onError(error.localizedDescription)
            }
        }
    }

    func connectTooltipEmbedding(channelId: String, rootBlock: String, messenger: FlutterBinaryMessenger) {
        guard let nubrickClient = self.nubrickClient else {
            return
        }
        let channel = FlutterMethodChannel(name: "Nativebrik/Embedding/\(channelId)", binaryMessenger: messenger)
        let accessor = nubrickClient.experiment.__do_not_use__render_uiview(
            json: rootBlock,
            onEvent: { event in
                channel.invokeMethod(ON_EVENT_METHOD, arguments: [
                    "name": event.name as Any?,
                    "deepLink": event.deepLink as Any?,
                    "payload": event.payload?.map({ prop in
                        return [
                            "name": prop.name,
                            "value": prop.value,
                            "type": prop.type
                        ]
                    }),
                ])
            },
            onNextTooltip: { pageId in
                channel.invokeMethod(ON_NEXT_TOOLTIP_METHOD, arguments: [
                    "pageId": pageId,
                ])
            },
            onDismiss: {
                channel.invokeMethod(ON_DISMISS_TOOLTIP_METHOD, arguments: nil)
            }
        )
        let embeedingEntity = EmbeddingEntity(
            uiview: accessor.view,
            channel: channel,
            accessor: accessor
        )
        self.embeddingMaps[channelId] = embeedingEntity
    }

    func callTooltipEmbeddingDispatch(channelId: String, event: String) {
        guard let entity = self.embeddingMaps[channelId] else {
            return
        }
        guard let accessor = entity.accessor else {
            return
        }
        do {
            try accessor.dispatch(event: event)
        } catch {}
    }

    func disconnectTooltipEmbedding(channelId: String) {
        self.embeddingMaps[channelId] = nil
    }

    // trigger
    func dispatch(name: String) {
        guard let nubrickClient = self.nubrickClient else {
            return
        }
        nubrickClient.experiment.dispatch(NubrickEvent(name))
    }

    /**
     * Sends crash events from Flutter.
     *
     * This method constructs a crash event and forwards it to the Nativebrik SDK for crash reporting
     * with platform set to "flutter".
     *
     * - Parameter exceptionsList: List of exception records from Flutter
     * - Parameter flutterSdkVersion: The Flutter SDK version
     * - Parameter severity: The severity level ("crash" or "warning")
     */
    func sendFlutterCrash(_ exceptionsList: [[String: Any?]], flutterSdkVersion: String?, severity: String?) {
        guard let nubrickClient = self.nubrickClient else {
            return
        }

        let exceptions = exceptionsList.compactMap { exceptionMap -> ExceptionRecord? in
            let type = exceptionMap["type"] as? String
            let message = exceptionMap["message"] as? String
            let callStacksList = exceptionMap["callStacks"] as? [[String: Any?]]

            let callStacks = callStacksList?.compactMap { frameMap -> StackFrame? in
                StackFrame(
                    fileName: frameMap["fileName"] as? String,
                    className: frameMap["className"] as? String,
                    methodName: frameMap["methodName"] as? String,
                    lineNumber: frameMap["lineNumber"] as? Int
                )
            }

            return ExceptionRecord(
                type: type,
                message: message,
                callStacks: callStacks
            )
        }

        if !exceptions.isEmpty {
            let crashEvent = TrackCrashEvent(
                exceptions: exceptions,
                platform: "flutter",
                flutterSdkVersion: flutterSdkVersion,
                severity: CrashSeverity.from(severity)
            )
            nubrickClient.experiment.sendFlutterCrash(crashEvent)
        }
    }
}

File: android/src/main/kotlin/app/nubrick/flutter/nubrick_flutter/Manager.kt
@file:OptIn(FlutterBridgeApi::class)

package app.nubrick.flutter.nubrick_flutter

import android.content.Context
import app.nubrick.nubrick.Config
import app.nubrick.nubrick.Event
import app.nubrick.nubrick.FlutterBridgeApi
import app.nubrick.nubrick.FlutterBridge
import app.nubrick.nubrick.NubrickSDK
import app.nubrick.nubrick.data.ExceptionRecord
import app.nubrick.nubrick.data.NotFoundException
import app.nubrick.nubrick.data.StackFrame
import app.nubrick.nubrick.data.CrashSeverity
import app.nubrick.nubrick.data.TrackCrashEvent
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import io.flutter.plugin.common.StandardMessageCodec
import android.view.View
import android.widget.LinearLayout
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import app.nubrick.nubrick.NubrickEvent
import app.nubrick.nubrick.NubrickSize
import app.nubrick.nubrick.NubrickProvider
import app.nubrick.nubrick.component.bridge.UIBlockActionBridge
import app.nubrick.nubrick.remoteconfig.RemoteConfigVariant
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

internal data class ConfigEntity(val variant: RemoteConfigVariant?, val experimentId: String?)

private fun nubrickSizeToMessage(size: NubrickSize): Map<String, Any?> {
    return when (size) {
        is NubrickSize.Fixed -> mapOf(
            "kind" to "fixed",
            "value" to size.value.toDouble(),
        )
        NubrickSize.Fill -> mapOf(
            "kind" to "fill",
        )
    }
}

internal class NubrickFlutterManager(
    private val binaryMessenger: BinaryMessenger,
    private val scope: CoroutineScope
) {
    private var embeddingMap: MutableMap<String, Any?> = mutableMapOf()
    private var eventBridgeViewMap: MutableMap<String, UIBlockActionBridge> = mutableMapOf()
    private var configMap: MutableMap<String, ConfigEntity> = mutableMapOf()

    fun initialize(
        context: Context,
        projectId: String,
        onEvent: (event: Event) -> Unit,
        onDispatch: (event: NubrickEvent) -> Unit,
        onTooltip: (data: String, experimentId: String) -> Unit
    ) {
        // Callbacks are passed at init to avoid missing events fired during initialization.
        NubrickSDK.initialize(
            context,
            Config(
                projectId,
                onEvent = onEvent,
                onDispatch = onDispatch,
            ),
            onTooltip = onTooltip
        )
        // initialize is idempotent — on subsequent calls it's a no-op, so update callbacks separately.
        FlutterBridge.updateCallbacks(
            onEvent = onEvent,
            onDispatch = onDispatch,
            onTooltip = onTooltip
        )
    }

    fun getUserId(): String? {
        return NubrickSDK.getUserId()
    }

    fun setUserProperties(properties: Map<String, Any>) {
        NubrickSDK.setUserProperties(properties)
    }

    fun getUserProperties(): Map<String, String>? {
        return NubrickSDK.getUserProperties()
    }

    // embedding
    fun connectEmbedding(channelId: String, experimentId: String, componentId: String? = null) {
        val methodChannel = MethodChannel(this.binaryMessenger, "Nubrick/Embedding/$channelId")
        scope.launch {
            val result = withContext(Dispatchers.IO) {
                FlutterBridge.connectEmbedding(experimentId, componentId)
            }
            result.onSuccess {
                embeddingMap[channelId] = it
                val size = FlutterBridge.computeInitialSize(it)
                methodChannel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, "completed")
                methodChannel.invokeMethod(EMBEDDING_INITIAL_SIZE_METHOD, mapOf(
                    "width" to nubrickSizeToMessage(size.first),
                    "height" to nubrickSizeToMessage(size.second),
                ))
            }.onFailure {
                when (it) {
                    is NotFoundException -> {
                        methodChannel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, "not-found")
                    }
                    else -> {
                        methodChannel.invokeMethod(EMBEDDING_PHASE_UPDATE_METHOD, "failed")
                    }
                }
            }
        }
    }

    fun disconnectEmbedding(channelId: String) {
        embeddingMap.remove(channelId)
    }

    @Composable
    fun Render(channelId: String, arguments: Any?, modifier: Modifier = Modifier) {
        if (channelId.isEmpty()) {
            return
        }
        val methodChannel = remember(channelId) {
            MethodChannel(this.binaryMessenger, "Nubrick/Embedding/$channelId")
        }
        val data = this.embeddingMap[channelId]
        val eventBridge = this.eventBridgeViewMap[channelId]
        FlutterBridge.render(
            modifier,
            arguments,
            data,
            onEvent = { event ->
                methodChannel.invokeMethod(ON_EVENT_METHOD, mapOf(
                    "name" to event.name,
                    "deepLink" to event.deepLink,
                    "payload" to event.payload?.map { prop ->
                        mapOf(
                            "name" to prop.name,
                            "value" to prop.value,
                            "type" to prop.type,
                        )
                    }
                ))
            },
            onNextTooltip = { pageId ->
                methodChannel.invokeMethod(ON_NEXT_TOOLTIP_METHOD, mapOf(
                    "pageId" to pageId,
                ))
            },
            onDismiss = {
                methodChannel.invokeMethod(ON_DISMISS_TOOLTIP_METHOD, null)
            },
            onSizeChange = { width, height ->
                methodChannel.invokeMethod(EMBEDDING_SIZE_UPDATE_METHOD, mapOf(
                    "width" to nubrickSizeToMessage(width),
                    "height" to nubrickSizeToMessage(height),
                ))
            },
            eventBridge = eventBridge
        )
    }

    @Composable
    fun RenderOverlay() {
        NubrickProvider {}
    }

    // remote config
    suspend fun connectRemoteConfig(channelId: String, experimentId: String): Result<String> {
        if (channelId.isEmpty()) {
            return Result.success("not-found")
        }
        this.configMap[channelId] = ConfigEntity(null, null)

        if (experimentId.isEmpty()) {
            return Result.success("not-found")
        }
        val remoteConfig = NubrickSDK.remoteConfig(experimentId).getOrElse {
            return Result.success("not-found")
        }
        val variant = remoteConfig.fetch().getOrElse {
            val status = when (it) {
                is NotFoundException -> "not-found"
                else -> "failed"
            }
            return Result.success(status)
        }
        if (this.configMap[channelId] != null) {
            this.configMap[channelId] = ConfigEntity(variant, variant.experimentId)
        }
        return Result.success("completed")
    }

    fun disconnectRemoteConfig(channelId: String) {
        this.configMap.remove(channelId)
    }

    fun getRemoteConfigValue(channelId: String, key: String): String? {
        if (channelId.isEmpty()) return null
        if (key.isEmpty()) return null
        val config = this.configMap[channelId] ?: return null
        val variant = config.variant ?: return null
        return variant.get(key)
    }

    fun connectEmbeddingInRemoteConfigValue(channelId: String, key: String, embeddingChannelId: String) {
        if (channelId.isEmpty()) return
        val config = this.configMap[channelId] ?: return
        val variant = config.variant ?: return
        val componentId = variant.get(key) ?: return
        val experimentId = config.experimentId ?: return
        this.connectEmbedding(embeddingChannelId, experimentId, componentId)
    }

    fun connectTooltipEmbedding(channelId: String, rootBlock: String) {
        if (channelId.isEmpty()) return
        embeddingMap[channelId] = rootBlock
        eventBridgeViewMap[channelId] = UIBlockActionBridge()
    }

    suspend fun callTooltipEmbeddingDispatch(channelId: String, event: String) {
        if (channelId.isEmpty() || event.isEmpty()) return
        val eventBridge = eventBridgeViewMap[channelId] ?: return
        eventBridge.dispatch(event)
    }

    fun disconnectTooltip(channelId: String) {
        if (channelId.isEmpty()) return
        embeddingMap.remove(channelId)
        eventBridgeViewMap.remove(channelId)
    }

    fun appendTooltipExperimentHistory(experimentId: String) {
        if (experimentId.isEmpty()) return
        NubrickSDK.appendTooltipExperimentHistory(experimentId)
    }

    fun dispatch(name: String) {
        NubrickSDK.dispatch(NubrickEvent(name))
    }

    /**
     * Records exceptions from Flutter.
     *
     * This method constructs a crash event and forwards it to the Nubrick SDK for crash reporting
     * with platform set to "flutter".
     *
     * @param exceptionsList List of exception records from Flutter
     * @param flutterSdkVersion The Flutter SDK version
     * @param severity The severity level ("crash" or "warning")
     */
    fun recordFlutterExceptions(exceptionsList: List<Map<String, Any?>>, flutterSdkVersion: String?, severity: String?) {
        try {
            val exceptions = exceptionsList.mapNotNull { exceptionMap ->
                try {
                    val type = exceptionMap["type"] as? String
                    val message = exceptionMap["message"] as? String
                    val callStacksList = exceptionMap["callStacks"] as? List<*>

                    val callStacks = callStacksList?.mapNotNull { frameMap ->
                        try {
                            val frame = frameMap as? Map<*, *>
                            StackFrame(
                                fileName = frame?.get("fileName") as? String,
                                className = frame?.get("className") as? String,
                                methodName = frame?.get("methodName") as? String,
                                lineNumber = (frame?.get("lineNumber") as? Number)?.toInt()
                            )
                        } catch (e: Exception) {
                            null
                        }
                    }

                    ExceptionRecord(
                        type = type,
                        message = message,
                        callStacks = callStacks
                    )
                } catch (e: Exception) {
                    null
                }
            }

            if (exceptions.isNotEmpty()) {
                val crashEvent = TrackCrashEvent(
                    exceptions = exceptions,
                    platform = "flutter",
                    flutterSdkVersion = flutterSdkVersion,
                    severity = CrashSeverity.from(severity)
                )
                NubrickSDK.sendFlutterCrash(crashEvent)
            }
        } catch (e: Exception) {
            // Silently fail to avoid causing crashes in error reporting
        }
    }
}

internal class OverlayViewFactory(private val manager: NubrickFlutterManager): PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        return OverlayView(context, manager)
    }
}

internal class OverlayView(context: Context, manager: NubrickFlutterManager): PlatformView {
    private val view: ComposeView

    override fun getView(): View {
        return view
    }

    override fun dispose() {}

    init {
        view = ComposeView(context).apply {
            setContent {
                manager.RenderOverlay()
            }
        }
    }
}

internal class NativeViewFactory(private val manager: NubrickFlutterManager): PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val creationParams = args as Map<*, *>?
        val channelId = creationParams?.get("channelId") as String
        val arguments = creationParams["arguments"]
        return NativeView(context, channelId, arguments, manager)
    }
}

internal class NativeView(context: Context, channelId: String, arguments: Any?, manager: NubrickFlutterManager): PlatformView {
    private val view: ComposeView

    override fun getView(): View {
        return view
    }

    override fun dispose() {}

    init {
        val param = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT)
        view = ComposeView(context).apply {
            setContent {
                manager.Render(channelId, arguments)
            }
            layoutParams = param
        }
    }
}
